File: advent_of_code_2023/day19/day19.ts
import {readFileSync} from "fs";

type Category = "x" | "m" | "a" | "s";

const MINIMUM: Part = {
    x: 1,
    m: 1,
    a: 1,
    s: 1
}
const MAXIMUM: Part = {
    x: 4001,
    m: 4001,
    a: 4001,
    s: 4001
}

interface Rule {
    accept: string;

    eval(part: Part): boolean;
    split(range: PartRange): [PartRange | null, PartRange | null];
}

class GreaterThan implements Rule {
    param: Category
    value: number
    accept: string;

    constructor(param: Category, value: number, accept: string) {
        this.param = param;
        this.value = value;
        this.accept = accept;
    }
    split(range: PartRange): [PartRange | null, PartRange | null] {
        if (this.eval(range.minPart)) {
            // The whole range true.
            return [range, null];
        } else if (this.eval(range.maxPart)) {
            // Partially true.
            let trueRange = range.copy();
            let falseRange = range.copy();
            trueRange.setMin(this.param, this.value + 1);
            falseRange.setMax(this.param, this.value + 1);
            
            return [trueRange, falseRange];
        } else {
            return [null, range];
        }
    }

    eval(part: Part): boolean {
        switch (this.param) {
            case "x":
                return part.x > this.value;
            case "m":
                return part.m > this.value;
            case "a":
                return part.a > this.value;
            case "s":
                return part.s > this.value;
        }
    }
}
class LessThan implements Rule {
    param: Category;
    value: number;
    accept: string;

    constructor(param: Category, value: number, accept: string) {
        this.param = param;
        this.value = value;
        this.accept = accept;
    }

    eval(part: Part): boolean {
        switch (this.param) {
            case "x":
                return part.x < this.value;
            case "m":
                return part.m < this.value;
            case "a":
                return part.a < this.value;
            case "s":
                return part.s < this.value;
        }
    }
    split(range: PartRange): [PartRange | null, PartRange | null] {
        if (this.eval(range.maxPart)) {
            // The whole range true.
            return [range, null];
        } else if (this.eval(range.minPart)) {
            // Partially true.
            let trueRange = range.copy();
            let falseRange = range.copy();
            trueRange.setMax(this.param, this.value);
            falseRange.setMin(this.param, this.value);
            return [trueRange, falseRange];
        } else {
            return [null, range];
        }
    }
}

class Workflow {

    rules: Rule[];
    else: string;

    constructor(rules: Rule[], _else: string) {
        this.rules = rules;
        this.else = _else;
    }
    eval(p: Part): string {
        for (let r of this.rules) {
            if (r.eval(p)) {
                return r.accept;
            }
        }
        return this.else;
    }
    getRangesAccepted(range: PartRange, workflowMap: Map<string, Workflow>, depth=0): number {
        if (depth > workflowMap.size) {
            return 0;
        }
        let currentRange = range;
        let accepted = 0;
        for (let r of this.rules) {
            let [t, f] = r.split(currentRange);
            if (t !== null) {
                let name = r.accept;
                if (name === 'A') {
                    accepted += t.getCombinations();
                } else if (name !== 'R') {
                    let next = workflowMap.get(name);
                    if (next) {
                        accepted += next.getRangesAccepted(t, workflowMap, depth + 1);
                    }
                }
            }
            if (f !== null) {
                currentRange = f;
            } else {
                return accepted;
            }
        }
        if (this.else === 'A') {
            return accepted + currentRange.getCombinations();
        } else if (this.else === 'R') {
            return accepted;
        } else {
            let next = workflowMap.get(this.else);
            if (next) {
                return accepted + next.getRangesAccepted(currentRange, workflowMap, depth + 1);
            }
        }
        return accepted;
    }
}

type Part = {
    [key in Category]: number;
};

class PartRange {
    minPart: Part;
    maxPart: Part; // Endpoint exclusive

    constructor(minPart: Part, maxPart: Part) {
        this.minPart = minPart;
        this.maxPart = maxPart;
    }

    split(rule: Rule): [PartRange | null, PartRange | null] {
        // Returns a pair of ranges that are true and false respectively
        return rule.split(this);
    }

    getCombinations(): number {
        return (this.maxPart.x - this.minPart.x) 
             * (this.maxPart.m - this.minPart.m)
             * (this.maxPart.a - this.minPart.a)
             * (this.maxPart.s - this.minPart.s);
    }
    private copyPart(p: Part): Part {
        return {
            x: p.x,
            m: p.m,
            a: p.a,
            s: p.s
        };
    }
    copy(): PartRange {
        return new PartRange(this.copyPart(this.minPart), this.copyPart(this.maxPart))
    }
    setMax(key: Category, n: number): void {
        this.maxPart[key] = Math.max(n, this.minPart[key]);
    }
    setMin(key: Category, n: number): void {
        this.minPart[key] = Math.min(n, this.maxPart[key]);
    }
}

function parseCategory(category: string): Category {
    switch (category) {
        case 'x':
            return 'x';
        case 'm':
            return 'm';
        case 'a':
            return 'a';
        case 's':
            return 's';
        default:
            throw new Error("Unknown Category: " + category);
    }
}
function parseRules(rules: string[]): [Rule[], string] {
    let elseRule = rules[rules.length - 1]
    let ruleArray: Rule[] = [];
    for (let i = 0; i < rules.length - 1; i++) {
        let match = rules[i].match(/(.+)(<|>)(\d+):(.+)/);
        if (match == null || match.length < 5) {
            throw new Error("Unable to parse rule:" + rules[i]);
        }
        let category = parseCategory(match[1])
        switch (match[2]) {
            case '<':
                ruleArray.push(new LessThan(category, parseInt(match[3]), match[4]));
                break;
            case '>': 
                ruleArray.push(new GreaterThan(category, parseInt(match[3]), match[4]));
                break;
            default:
                throw new Error("Unknown operator in rule:" + rules[i]);
        }

    }
    return [ruleArray, elseRule];
}

function parseWorkflow(line: string): [string, Workflow] {
    let match = line.match(/(.+){(.+)}/);
    if (match === null || match.length < 3) {
        throw new Error("Unable to parse Workflow: " + line);
    }
    let name = match[1];
    try {
        let [rules, _else] = parseRules(match[2].split(/,/));
        return [name, new Workflow(rules, _else)];
    } catch {
        throw new Error("Unable to parse Workflow: " + line);
    }
}
function parseWorkflows(workflowLines: string[]): Map<string, Workflow> {
    let workflowMap = new Map<string, Workflow>();
    for (let str of workflowLines) {
        if (str.length == 0) {
            continue;
        }
        let [name, workflow] = parseWorkflow(str);
        workflowMap.set(name, workflow);
    }
    return workflowMap;
}
function parseParts(partLines: string[]): Part[] {
    let parts: Part[] = [];
    for (let line of partLines) {
        if (line.length == 0) {
            continue;
        }
        let match = line.match(/{x=(\d+),m=(\d+),a=(\d+),s=(\d+)}/);
        if (match === null || match.length < 5) {
            throw new Error("Unable to parse Part: " + line);
        }
        parts.push({
            x: parseInt(match[1]),
            m: parseInt(match[2]),
            a: parseInt(match[3]),
            s: parseInt(match[4])
        })
    }
    return parts;
}

function isAccepted(p: Part, workflowMap: Map<string, Workflow>): boolean {
    let currentWorkflow = 'in';
    let i = 0
    while (i < workflowMap.size) {
        let w = workflowMap.get(currentWorkflow);
        if (w) {
            let output = w.eval(p);
            if (output == 'R') {
                return false;
            } else if (output == 'A') {
                return true;
            } else {
                currentWorkflow = output;
            }
        } else {
            return false;
        }
        i++;
    }
    return false;

}
function partSum(p: Part): number {
    return p.x + p.m + p.a + p.s;
}

function solvePart2(range: PartRange, workflow: string, 
                    workflowMap: Map<string, Workflow>): number {
    let w = workflowMap.get(workflow);
    if (w === undefined) {
        return 0;
    }
    return w.getRangesAccepted(range, workflowMap);
    

}

function solve(lines: string[]): void {
    let split = lines.findIndex((v) => v.length == 0);
    let workflowMap = parseWorkflows(lines.slice(0, split));
    let parts = parseParts(lines.slice(split + 1));

    let sum = 0;
    for (let p of parts) {
        if (isAccepted(p, workflowMap)) {
            sum += partSum(p);
        }
    }
    console.log("Part 1:", sum);

    let range = new PartRange(MINIMUM, MAXIMUM);
    console.log("Part 2:", solvePart2(range, 'in', workflowMap));

}

function main() {
    let lines: string[] = readFileSync("input19.txt", "utf-8")
                            .split(/\n/)
                            .map(x => x.trim())
    console.time("Time taken");
    solve(lines);
    console.timeEnd("Time taken");
}

main();

File: advent_of_code_2023/day19/day19.py
from collections import deque

def arrayise(filename):
    array = []
    f = open(filename, "r")
    for line in f:
        array.append(line.strip())
    return array

class Rule:
    def __init__(self, rules):
        self.else_function = rules[-1]
        self.comparisons = [] # (param, operator, value, output if true)
        for r in rules[:-1]:
            comparison, output = r.split(':')
            if comparison[1] == '>':
                param, num = comparison.split('>')
                self.comparisons.append((param, '>', int(num), output))
            else:
                param, num = comparison.split('<')
                self.comparisons.append((param, '<', int(num), output))
    def apply(self, part):
        for cmp in self.comparisons:
            param = part.x
            if cmp[0] == 'm':
                param = part.m
            elif cmp[0] == 'a':
                param = part.a
            elif cmp[0] == 's':
                param = part.s
            
            if cmp[1] == '<':
                if (param < cmp[2]):
                    return cmp[3]
            elif cmp[1] == '>':
                if (param > cmp[2]):
                    return cmp[3]
        return self.else_function
                

class Part:
    def __init__(self, part):
        tokens = part[1:-1].split(',')

        self.x = int(tokens[0].split('=')[1])
        self.m = int(tokens[1].split('=')[1])
        self.a = int(tokens[2].split('=')[1])
        self.s = int(tokens[3].split('=')[1])

class Range:
    def __init__(self, xmin, xmax, mmin, mmax, amin, amax, smin, smax):

        self.xmin = xmin
        self.xmax = xmax
        self.mmin = mmin
        self.mmax = mmax
        self.amin = amin
        self.amax = amax
        self.smin = smin
        self.smax = smax
    
    def separate(self, cmp):
        # Comparison: (param, operator, value, output if true)
        mint = self.xmin
        maxt = self.xmax
        if cmp[0] == 'm':
            mint = self.mmin
            maxt = self.mmax
        elif cmp[0] == 'a':
            mint = self.amin
            maxt = self.amax
        elif cmp[0] == 's':
            mint = self.smin
            maxt = self.smax
        minf = mint
        maxf = maxt

        t_notexists = False
        f_notexists = False
        if cmp[1] == '<':
            mint = min(mint, cmp[2])
            maxt = min(maxt, cmp[2] - 1)
            if (mint == cmp[2]):
                t_notexists = True
            
            minf = max(minf, cmp[2])
            maxf = max(maxf, cmp[2] - 1)
            if (maxf == cmp[2] - 1):
                f_notexists = True
        elif cmp[1] == '>':
            mint = max(mint, cmp[2] + 1)
            maxt = max(maxt, cmp[2])
            if (maxt == cmp[2]):
                t_notexists = True

            minf = min(minf, cmp[2] + 1)
            maxf = min(maxf, cmp[2])
            if (minf == cmp[2] + 1):
                f_notexists = True
        
        r1 = self.copy()
        r2 = self.copy()
        if (t_notexists):
            r1 = None
        else:
            if cmp[0] == 'm':
                r1.mmin = mint
                r1.mmax = maxt
            elif cmp[0] == 'a':
                r1.amin = mint
                r1.amax = maxt
            elif cmp[0] == 's':
                r1.smin = mint
                r1.smax = maxt
            else:
                r1.xmin = mint
                r1.xmax = maxt
        if (f_notexists):
            r2 = None
        else:
            if cmp[0] == 'm':
                r2.mmin = minf
                r2.mmax = maxf
            elif cmp[0] == 'a':
                r2.amin = minf
                r2.amax = maxf
            elif cmp[0] == 's':
                r2.smin = minf
                r2.smax = maxf
            else:
                r2.xmin = minf
                r2.xmax = maxf
        return [r1, r2]
    def copy(self):
        return Range(self.xmin, self.xmax, self.mmin, self.mmax, self.amin, self.amax, self.smin, self.smax)
    def combos(self):
        return (self.xmax - self.xmin + 1)*(self.mmax - self.mmin + 1)*(self.amax - self.amin + 1) *(self.smax - self.smin + 1)

def solve_part2_helper(r, name, functions):
    if (r == None):
        return 0
    elif (name == 'R'):
        return 0
    elif (name == 'A'):
        return r.combos()
    func = functions[name]
    summation = 0
    curr_range = r
    for rule in func.comparisons:
        true_range, false_range = curr_range.separate(rule)
        summation += solve_part2_helper(true_range, rule[3], functions)
        curr_range = false_range
        if (curr_range == None):
            break
    return summation + solve_part2_helper(curr_range, func.else_function, functions)

def solve_part2(functions):
    r = Range(1, 4000, 1, 4000, 1, 4000, 1, 4000)
    print(solve_part2_helper(r, 'in', functions))


def solve(lines):
    rules = []
    parts = []
    currarr = rules

    
    for i, line in enumerate(lines):
        if line == '':
            currarr = parts
        else:
            currarr.append(line)
    
    functions = dict()
    parts_obj = []
    for r in rules:
        name, func = r.split('{')
        func = func[:-1]
        cmps = func.split(',')
        
        functions[name] = Rule(cmps)
    for p in parts:
        parts_obj.append(Part(p))
    
    summation = 0
    for p in parts_obj:
        rule = 'in'
        while True:
            if rule == 'R':
                break
            if rule == 'A':
                summation += p.x + p.m + p.a + p.s
                break
            rule = functions[rule].apply(p)
    print(summation)

    solve_part2(functions)

if __name__ == '__main__':
    filename = "input19.txt"
    arr = arrayise(filename)
    solve(arr)